### PR TITLE
CI: GitHub action for checks (linting, docstrings...)

### PR DESCRIPTION
**Pull request by datapythonista**
xref #29472

### CODE REVIEW

**Comment by jbrockmendel** [comment]
Is this supposed to take the place of something currently in travis/azure?  or totally new?

**Comment by datapythonista** [comment]
This is replicating part of the `Web_and_Docs` build in azure. I expected this PR to create a new check, next to the travis and azure builds, to see how it looks like. But seems like it needs to be committed to run (or I am missing something).

I guess in some time we'll want to replace both azure pipelines and travis with GitHub actions, but it's early to start discussing about it.

**Comment by datapythonista** [comment]
@pandas-dev/pandas-core here you can see how GitHub actions will look like:
- https://github.com/pandas-dev/pandas/pull/29546/checks?check_run_id=306452705

The main advantage over azure-pipelines is that clicking on `Details` takes you directly to the list of steps/errors. I also find the yaml syntax of actions a bit clearer than pipelines.

If we're happy to move forward, we can merge this, and leave the `Checks` build for a while in both systems, until we're confident with actions, and then we keep moving other things.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        export PATH=$HOME/miniconda3/bin:$PATH
+        source activate pandas-dev
+        ci/code_checks.sh docstrings
+      if: true
+
+    - name: Typing validation
+      run: |
+        export PATH=$HOME/miniconda3/bin:$PATH
+        source activate pandas-dev
+        ci/code_checks.sh docstrings
```
Comment (jreback) [inline .github/workflows/ci.yml]: is this right? do we have a separate code checks for typing

**Review by jreback — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 1 above)
Comment (datapythonista) [reply .github/workflows/ci.yml]: Good catch. I guess I copy-pasted incorrectly while transforming the yaml.

Fixed. I also remove the activate action here, which is not needed anymore now that actions is already active, and new actions are executed in PRs.

**Comment by jreback** [comment]
lgtm. can you rebase

**Comment by datapythonista** [comment]
@jreback rebased and all green

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks

**Comment by gfyoung** [comment]
https://github.com/pandas-dev/pandas/runs/306852441#step:3:36

@datapythonista @jreback : I think this was merged a little too early.  What's going on here?

**Comment by jreback** [comment]
ahh it’s not actually failing

thanks for the catch @gfyoung 

@datapythonista can u follow up

**Comment by datapythonista** [comment]
Wow, it's really amazing that we can link to a specific line in the log. :) I don't think we had this before, right?

Thanks for the heads up @gfyoung. I don't quite understand what's going on, but let me have a look.

**Comment by gfyoung** [comment]
> Wow, it's really amazing that we can link to a specific line in the log.

Travis CI has had this functionality for quite some time.  Why Azure did not escapes me 🙂.

**Comment by wesm** [comment]
Just to deliver some feedback we've just developed an extensive GHA configuration in Apache Arrow and we are pleased so far (see https://github.com/apache/arrow/tree/master/.github/workflows), if you have any questions @kszucs would be a good person to talk to, there's probably plenty we can learn from our respective experiences

**Comment by datapythonista** [comment]
Thanks for the info @wesm. Looks quite interesting, need to check in more detail.

I just opened a PR for a draft of our CI in actions: #29715

@kszucs happy to discuss and share experiences.

**Comment by kszucs** [comment]
@datapythonista we keep our linux builds in docker containers, which is CI agnostic, makes local reproducibility possible and speeds up the builds in case of many dependencies. I'll try to take a closer look on your current CI scripts and your PR in the upcoming days.

**Pull request merged**